### PR TITLE
fix: legacy gas miss match and not preserved after change

### PR DESCRIPTION
**Pull request by blackdevelopa**
## **Description**

Whenever I trigger a legacy transaction from a dapp, I can see how the Estimated gas fees does not match the gas details (if I open the Edit gas). Furthermore if I change the suggested gas, I can see how the value is not changed and tx is submitted with the suggested gas value instead of the one I set. [See here](https://github.com/MetaMask/metamask-mobile/pull/7035#issuecomment-1717118939)

_Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:_
_1. What is the reason for the change?_
_2. What is the improvement/solution?_

## **Manual testing steps**
Test for both 1559 supported network and legacy network. For legacy network, use ganache and check that `--hardfork berlin` or use Palm. For 1559, a testnet like Georli should suffice

Setup legacy network using 
_1. Step1:_ Setup wallet and import the right key and account
_3. Step3:_ Check that transactions are submitted as expected.

### **Before**

https://github.com/MetaMask/metamask-mobile/assets/29962968/b816bfb0-fe78-4b86-81a4-0e554cfd38cc

https://github.com/MetaMask/metamask-mobile/assets/29962968/c7d7b889-2fac-4560-8896-e1b330b42f13

### **After**
Send token: https://recordit.co/KCujgfnwOO
Send legacy: http://recordit.co/QOSHpcmrMS
Create a token: http://recordit.co/ygpx2T3Wsw
Token approve: http://recordit.co/C4ItxYxWCB
Send legacy txn from Georli: http://recordit.co/TUmaOp9vHl

## **Related issues**

_Fixes #7290 

Bitrise build: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/35e839a1-6d20-491d-bfea-6956c207776a

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/components/UI/EditGasFeeLegacyUpdate/EditGasFeeLegacyUpdate.test.tsx [hunk 1]
```diff
@@ -17,8 +17,7 @@ const mockInitialState = (txnType = 'none') => ({
 
 const selectedGasObjectForFeeMarket = {
   legacyGasLimit: undefined,
-  suggestedGasPrice: undefined,
-  suggestedMaxFeePerGas: '10',
+  suggestedGasPrice: '10',
 };
```
Comment (jpuri) [inline app/components/UI/EditGasFeeLegacyUpdate/EditGasFeeLegacyUpdate.test.tsx:21]: Above looks confusing as 1559 does not have `gasPrice` but it has `suggestedMaxFeePerGas`.

File: app/core/GasPolling/GasPolling.ts [hunk 2]
```diff
@@ -207,19 +207,30 @@ export const useGasTransaction = ({
...
-      : gasObjectLegacy?.suggestedGasPrice ||
-        gasFeeEstimates?.gasPrice ||
-        gasFeeEstimates?.low;
+  let suggestedGasPrice;
+
+  if (gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET) {
+    suggestedGasPrice = gasObjectLegacy?.suggestedGasPrice;
+  } else {
+    suggestedGasPrice = gasFeeEstimates?.gasPrice || gasFeeEstimates?.low;
+  }
```
Comment (jpuri) [inline app/core/GasPolling/GasPolling.ts:222]: I am bit confused why gasPrice estimate is required for 1559 transaction ?

**Comment by jpuri** [comment]
@blackdevelopa : code changes look good, I left some small queries can you plz check.

File: app/core/GasPolling/GasPolling.ts (hunk 2 above)
Comment (blackdevelopa) [reply app/core/GasPolling/GasPolling.ts:222]: Hey @jpuri, great question.

We still support calling legacy transaction (type === 0x0) from a 1559 supported network. For this, we rely on `gasFeeEstimates?.gasPrice || gasFeeEstimates?.low` values for suggestedGasPrice. However, if either of the gas limit or gas price from the EditModal is adjusted, we save the new values in [gasObjectLegacy](https://github.com/MetaMask/metamask-mobile/blob/f09fd7728d33b6dec7abd6ef075c46403aa5762e/app/core/GasPolling/types.ts#L135) which is then used to [recalculate](https://github.com/MetaMask/metamask-mobile/blob/f09fd7728d33b6dec7abd6ef075c46403aa5762e/app/core/GasPolling/GasPolling.ts#L149) the new gas fees.

File: app/components/UI/EditGasFeeLegacyUpdate/EditGasFeeLegacyUpdate.test.tsx (hunk 1 above)
Comment (blackdevelopa) [reply app/components/UI/EditGasFeeLegacyUpdate/EditGasFeeLegacyUpdate.test.tsx:21]: Ah yes, this is valid. 1559 doesn't have gasPrice. Updated https://github.com/MetaMask/metamask-mobile/pull/7305/commits/e23db369c3dab05e728489148c02841341dee76b

File: app/components/UI/TransactionEditor/index.js [hunk 3]
```diff
@@ -680,21 +680,66 @@ class TransactionEditor extends PureComponent {
...
+    totalMaxHex: legacyGasTransaction?.totalHex,
+
+    // These values are updated to be able to submit to the network
+    suggestedMaxFeePerGas: legacyGasTransaction?.suggestedGasPrice,
+    suggestedMaxFeePerGasHex: legacyGasTransaction?.suggestedGasPriceHex,
+    suggestedMaxPriorityFeePerGas: legacyGasTransaction?.suggestedGasPrice,
+    suggestedMaxPriorityFeePerGasHex:
+      legacyGasTransaction?.suggestedGasPriceHex,
+  });
+
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/UI/TransactionEditor/index.js:709]: Sure! These values are derived from [parseTransactionLegacy](https://github.com/MetaMask/metamask-mobile/blob/c865d3c216eecbadbfcdbc59cba67d5dbaf112f3/app/util/transactions/index.js#L1191) and saved to `legacyGasTransaction`.

**Review by jpuri — DISMISSED** [review]

**Comment by seaona** [comment]
For some reason, Sepolia is still showing the gas missmatch. It's interesting because it's indeed fixed in Goerli, so we should wonder why both networks behave differently, since in theory , they should be the same :thinking: 

https://github.com/MetaMask/metamask-mobile/assets/54408225/53eb3687-9161-4ff1-9d0b-1ae64708c666

**Review by jpuri — APPROVED** [review]

**Comment by blackdevelopa** [comment]
Hey @seaona thanks for the QA feedback. Here's an update


https://github.com/MetaMask/metamask-mobile/assets/29962968/f26e6946-601e-4bbd-a7ad-2ddf9e5d8cac

**Comment by seaona** [comment]
thank you @blackdevelopa I also see the issue fixed


https://github.com/MetaMask/metamask-mobile/assets/54408225/ed8a34cf-bd7e-4b3e-a29f-e3912e948203

**Review by segun — APPROVED** [review]

**Pull request merged**